### DISCUSSION TITLE
cam_test camera_center_diff

### DISCUSSION THREAD

**Discussion by jlaura** [NeoGeographyToolkit/StereoPipeline#388] [Q&A] [answered]
The [`cam_test` documentation](https://stereopipeline.readthedocs.io/en/latest/tools/cam_test.html) states:

> It find(s) the camera center and ray direction at a set of sampled pixels for both cameras and checks how they differ. (typo in 'find').

Are the reported values the difference in the sensor location in x,y,z space w.r.t. some datum or the signed(?) deltas w.r.t. the cam1 location?

**Comment by oleg-alexandrov** [comment, marked as answer]
Jay,

Thank you. I fixed the typo. I also expanded the doc a bit: It now says:

This tool compares two camera models for the same image, or a camera
model against itself. It finds the camera center and ray direction at a
set of sampled pixels for both cameras and checks how they
differ. Then, it projects pixels to the ground using the first camera
and back-projects the resulting ECEF points into the second camera, and
estimates the pixel discrepancy.

So, three comparisons are done (1) camera center, this is in units of meter
(2) ray direction, (3) projection into camera, with the latter in units of
pixel. It does not compute signed deltas, only norm of of those 3D and 2D
vectors (for camera centers and pixels, respectively). Let me know if
there's a need for more detail.

I did not include a sample output for this tool in the doc, as being too
verbose, but for the record, it is the following, and it is also suggestive
about what the tool does:

cam1 to cam2 camera direction diff norm
Min:    8.67632e-10
Median: 1.7293e-07
Max:    9.0646e-07

cam1 to cam2 camera center diff (meters)
Min:    3.79606e-06
Median: 5.1208e-05
Max:    0.308461

cam1 to cam2 pixel diff
Min:    4.03812e-05
Median: 0.0122446
Max:    0.0465756

cam2 to cam1 pixel diff
Min:    4.83225e-05
Median: 0.0121175
Max:    0.0466092

**Reply by jlaura** [reply]
@oleg-alexandrov That's terrific. Thanks! I have been using this also as an easy method to see what the adjustments look like in sensor space for images data pre- and post-adjustment. So, a side benefit because the tool is so convenient to use! Many thanks for developing and releasing it.

**Comment by oleg-alexandrov** [comment]
And I am not sure if I actually answered your questions. All xyz
measurements of sensor location are in ECEF, which kind of seems like the
logical coordinate system for this tool as those are orbital data rather
than on planet surface.

**Comment by jlaura** [comment]
@oleg-alexandrov Both of these are terrific answers. I'll make the first one, the second one answers my question as well.
